node/libs/concurrency/src/sync/prunable_queue.rs

```diff
@@ -0,0 +1,148 @@
+        queue,
+        has_items_send: send,
+    });
+
+    let sender = Sender {
+        shared: shared.clone(),
+        pruning_predicate: pruning_predicate,
+    };
+
+    let receiver = Receiver {
+        shared: shared.clone(),
+        has_items_recv: recv,
+    };
+
+    return (sender, receiver);
+}
+
+pub struct Shared<T, U> {
+    queue: Mutex<VecDeque<(T, oneshot::Sender<U>)>>,
+    has_items_send: watch::Sender<bool>,
+}
+
+pub struct Sender<T, U> {
+    shared: Arc<Shared<T, U>>,
+    pruning_predicate: Box<dyn Sync + Send + Fn(&T, &T) -> bool>,
+}
+
+impl<T, U> Sender<T, U> {
+    pub async fn enqueue(&self, item: T) -> oneshot::Receiver<U> {
+        // Create oneshot channel for returning result asynchronously.
+        let (res_send, res_recv) = oneshot::channel();
+
+        let mut queue = self.shared.queue.lock().await;
+        queue.retain(|existing_item| !(self.pruning_predicate)(&existing_item.0, &item));
+        queue.push_back((item, res_send));
+
+        // Ignore sending error.
+        let _ = self.shared.has_items_send.send(true);
+
+        res_recv
+    }
+}
+
+pub struct Receiver<T, U> {
+    shared: Arc<Shared<T, U>>,
+    has_items_recv: watch::Receiver<bool>,
+}
+
+impl<T, U> Receiver<T, U> {
+    pub async fn dequeue(&mut self, ctx: &Ctx) -> ctx::OrCanceled<(T, oneshot::Sender<U>)> {
+        sync::wait_for(ctx, &mut self.has_items_recv, |has_items| *has_items).await?;
+        let mut queue = self.shared.queue.lock().await;
+        // `None` is unexpected because we waited for new items.
+        let item = queue.pop_front().unwrap();
+
+        if queue.len() == 0 {
+            // Send error is unexpected because `self` holds the receiver.
+            self.shared.has_items_send.send(false).unwrap();
+        }
+
+        Ok(item)
+    }
+}
+
+// Test scenario:
+// Enqueue two sets of 0..1000 items, in conjunction, while pruning
+// so that only one 0..1000 set is expected to remain in queue.
+// Then, dequeue to assert the queue's content.
+#[tokio::test]
+async fn test_prunable_queue() {
+    use tokio::time::{timeout, Duration};
+
+    #[derive(Debug, Clone)]
+    struct ItemType(usize, usize);
+
+    let ctx = ctx::test_root(&ctx::RealClock);
+
+    let (sender, mut receiver) = new(Box::new(|a: &ItemType, b: &ItemType| a.0 != b.0));
+
+    let sender1 = Arc::new(sender);
+    let sender2 = sender1.clone();
+
+    let handle1 = tokio::spawn(async move {
+        let set = 1;
+        let items = (0..1000).map(|i| ItemType(i, set));
+        for item in items {
+            let _ = sender1.enqueue(item).await;
+            tokio::task::yield_now().await;
+        }
+    });
+    let handle2 = tokio::spawn(async move {
+        let set = 2;
+        let items = (0..1000).map(|i| ItemType(i, set));
+        for item in items {
+            let _ = sender2.enqueue(item).await;
+            tokio::task::yield_now().await;
+        }
+    });
+    tokio::try_join!(handle1, handle2).unwrap();
```

Comment:
please use structured concurrency for writing tests, unless infeasible.

Reply:
also note that using enqueue/dequeue like this requires guaranteeing cancel safety. Cancel safety when accepting ctx as an argument makes little sense (cancel safety: dropping a future is ok; accepting a ctx: to cancel, cancel the context and wait for the future to complete by itself)

Reply:
Can you clarify what you mean in the context of the test code? are you referring to being cancel-safe in the loops?

Regarding `recv()` (and unrelated to this test code)—it accepts `ctx` as an argument for cancellation watch. Watching on the loop level isn't sufficient, because `recv()` can hang indefinitely.

Reply:
Test code changes here: https://github.com/matter-labs/era-consensus/pull/56/commits/020f25b2049f9f4bb2dbbddc9a0e876eadaf5389

Reply:
Cancel safety is defined here: https://docs.rs/tokio/latest/tokio/macro.select.html#cancellation-safety.
tokio::timeout() requires the passed future to be cancel safe, because it drops it in case of timeout (https://docs.rs/tokio/latest/tokio/time/fn.timeout.html#cancellation). Additionally for this assert to work as intended, the `Receiver::recv(ctx)` here would need to guarantee that it completes immediately in case the queue is non-empty (which is not documented as of now). `Receiver::recv(ctx)` calls  `sync::wait_for` which doesn't provide such guarantee, so it would be hard to provide this guarantee. In general, none of the methods in concurrency crate do, as it is a very fragile property.

Reply:
I understand the problem. Follow-up here: https://github.com/matter-labs/era-consensus/pull/56#discussion_r1462277267

Reply:
I understand the problem, assuming you only refer to the timeout in the test code. Follow-up here: https://github.com/matter-labs/era-consensus/pull/56#discussion_r1462277267